BUG: fix infer frequency for business daily

AdamGleave opened the pull request:
Closes #16624 

Had to update a couple of tests, test_timedelta and test_business_daily, which appear to rely on the old incorrect behaviour. Added a new regression test, test_business_daily_look_alike, for the bug.

Modest performance impact: for 50,000-long business daily index, infer_freq goes from 2.54 ms to 3.91 ms. There should be no performance impact in most cases for non-business daily index, as the initial day_deltas check will throw out most of these cases.

Commit:
Fixes GH16624

jreback commented on pandas/tseries/frequencies.py:
```diff
@@ -1012,6 +1011,19 @@ def _get_monthly_rule(self):
         return {'cs': 'MS', 'bs': 'BMS',
                 'ce': 'M', 'be': 'BM'}.get(pos_check)
 
+    WORKING_DAY_SHIFTS = set([(0, 1), (1, 2), (2, 3), (3, 4), (4, 0)])
```
what is this for?

AdamGleave replied:
Unused code from a previous iteration, removed in latest commit

jreback reviewed (changes requested):
can you add an asv for this?

Commit:
Remove unused code

AdamGleave commented:
ASV report 
[attached](https://github.com/pandas-dev/pandas/files/1068792/asv_report.txt), indicates that no benchmarks are significantly changed.

TomAugspurger commented:
@jreback are you OK with the changes here? LGTM on a quick skim.

jreback commented on pandas/tests/indexes/timedeltas/test_timedelta.py:
```diff
@@ -567,7 +567,7 @@ class TestSlicing(object):
     def test_timedelta(self):
         # this is valid too
-        index = date_range('1/1/2000', periods=50, freq='B')
+        index = date_range('1/1/2000', periods=50, freq='D')
```
can u also test with B

jreback reviewed (approved):
small test comment
might be worth it to add an asv

jreback commented:
can you update

Commit:
TST: timedelta arithmetic on DatetimeIndex with 'B' and 'D' frequency

Commit:
ASV: test performance of infer_freq on business daily index

Commit:
ASV: more nuanced tests of infer_freq

jreback commented on doc/source/whatsnew/v0.20.3.txt:
```diff
@@ -66,7 +66,7 @@ Plotting
 ^^^^^^^^^^^^^^^^^^^^^^^^
 
-
+- Bug in ``infer_freq`` causing indices with 2-day gaps during the working week to be wrongly inferred as business daily (:issue:`16624`)
```
move to 0.21.0

jreback commented on pandas/tests/indexes/timedeltas/test_timedelta.py:
```diff
@@ -567,26 +567,34 @@ class TestSlicing(object):
-        result4 = DatetimeIndex(s - pd.offsets.Hour(1))
-        tm.assert_index_equal(result1, result4)
-        tm.assert_index_equal(result2, result3)
+        for freq in ['D', 'B']:
```
instead of doing a loop, use parametrize

jreback commented on pandas/tests/tseries/test_frequencies.py:
```diff
@@ -504,9 +504,14 @@ def test_raise_if_too_few(self):
         assert frequencies.infer_freq(index) == 'B'
 
+    def test_business_daily_look_alike(self):
+        # 'weekend' (2-day gap) in wrong place
```
add the issue number as a comment

jreback commented on pandas/tseries/frequencies.py:
```diff
@@ -1012,6 +1011,17 @@ def _get_monthly_rule(self):
                 'ce': 'M', 'be': 'BM'}.get(pos_check)
 
+    def _is_business_daily(self):
+        if self.day_deltas != [1, 3]:  # quick check: cannot be business daily
```
put the comment on the line above (causes lint failure as line too long)

jreback commented on pandas/tseries/frequencies.py:
```diff
@@ -1012,6 +1011,17 @@ def _get_monthly_rule(self):
+    def _is_business_daily(self):
+        if self.day_deltas != [1, 3]:  # quick check: cannot be business daily
+            return False
+        # probably business daily, but need to confirm
```
add blank line before comment

jreback commented on pandas/tseries/frequencies.py:
```diff
@@ -1012,6 +1011,17 @@ def _get_monthly_rule(self):
+            return False
+        # probably business daily, but need to confirm
+        first_weekday = self.index[0].weekday()
+        shifts = np.diff(np.asarray(self.index).view('i8'))
```
use ``np.diff(self.index.asi8)``

jreback reviewed (changes requested):
ping on green.

Commit:
Address jreback's review

whatsnew: bump from 0.20.3 to 0.21.0
test_timedelta: parametrize, strengthen index check
test_frequencies: add issue number
frequencies: formatting, np.asarray -> asi8

jreback commented:
looks like a tiny lint error.

Commit:
TST: fix lint error

jreback commented:
thanks!
